falcon/response.py

```diff
@@ -136,6 +145,82 @@ def set_stream(self, stream, stream_len):
         self.stream = stream
         self.stream_len = stream_len
 
+    def set_cookie(self, name, value, expires=None, max_age=None,
+                   domain=None, path=None, secure=True, httponly=True):
+        """Set a response cookie.
+
+        Args:
+            name (str):
+                Cookie name
+            value (str):
+                Cookie value
+            expires (datetime): Defines when the cookie should expire.
+                the default is to expire when the browser is closed.
+            max_age (int): The Max-Age attribute defines the lifetime of the
+                cookie, in seconds.  The delta-seconds value is a decimal non-
+                negative integer.  After delta-seconds seconds elapse,
+                the client should discard the cookie.
+            domain (str): The Domain attribute specifies the domain for
+                which the cookie is valid.
+                An explicitly specified domain must always start with a dot.
+                A value of 0 means the cookie should be discarded immediately.
+            path (str): The Path attribute specifies the subset of URLs to
+                which this cookie applies.
+            secure (bool) (default: True): The Secure attribute directs the
```

Comment:
We could easily add some magic that refuses to send a response if it contains secure cookies, but is attempted over a non-secure connection.

I think this has to go just before `start_response`, as you can have custom response types and middlewares that modifies the response. So I think the most robust point for checking this is after all possible modifications of the response.

But I think this could lead to some very annoying situations for a developer, that would be horrible to debug.

It also means that the default `secure=True` suddenly becomes impractical.

If we want to protect the developer against a bad configuration, I think it would be better to do it for _all_ requests.
Yes, the cookie might contain some session priviledges, but who's to say that the response without a cookie isn't equally important (especially in falcon, where most things are probably REST, and API keys are flying around all over the place).
So, simply refusing to send _any_ response over an insecure connection if the magical "be secure, please"-flag is turned on seems like a better "be secure by default" solution.

(Personal opinion: I don't think it is the responsibility of Falcon (read: application) to ensure that the transport to a user is secure. Whoever sets up the LB, webserver, etc. should know the desired configuration, secure or not.)